### PR TITLE
[Blazor] Correct case of *.js paths and fail fast w/o NodeJS

### PR DESCRIPTION
**Pull request by dougbu**
- #29944
- avoid issues when building on a case-sensitive file system
  - support varying `$(Configuration)` casing
- support building Components.Server w/o NodeJS even in Debug builds

### CODE REVIEW

**Review by dougbu — COMMENTED** [review]
Found a couple more case-related problems building in Ubuntu under WSL2. Corner cases but&hellip;

File: src/Components/WebAssembly/WebAssembly/src/Microsoft.AspNetCore.Components.WebAssembly.csproj [hunk 1]
```diff
@@ -44,11 +44,21 @@
...
+        Condition=" '$(Configuration)' != 'Debug' ">..\..\..\Web.JS\dist\Release\blazor.webassembly.js</BlazorWebAssemblyJSFile>
   </PropertyGroup>
 
   <ItemGroup>
     <Content Include="$(BlazorWebAssemblyJSFile)" Pack="true" PackagePath="build\$(DefaultNetCoreTargetFramework)\" LinkBase="build\$(DefaultNetCoreTargetFramework)\" />
     <Content Include="build\$(DefaultNetCoreTargetFramework)\*.props" Pack="true" PackagePath="build\$(DefaultNetCoreTargetFramework)\" />
   </ItemGroup>
+
+  <!-- blazor.webassembly.js should exist after Microsoft.AspNetCore.Components.Web.JS.npmproj builds. -->
+  <Target Name="_CheckBlazorServerJSPath" BeforeTargets="GenerateNuspec" DependsOnTargets="ResolveProjectReferences">
```
Comment (dougbu) [inline src/Components/WebAssembly/WebAssembly/src/Microsoft.AspNetCore.Components.WebAssembly.csproj:60]: This target needs refinement. It executes in many build jobs using `-p:OnlyPackPlatformSpecificPackages=true` and fails the build. Will try checking `$(IsPackable)`&hellip

**Comment by dougbu** [comment]
FYI in a local Ubuntu run of `./build.sh --all --no-build-native --no-build-nodejs`, see expected, newly-refined errors:
``` text
...src/aspnetcore/src/Components/Directory.Build.targets(30,5): error : '...src/aspnetcore/src/Components/Web.JS/dist/Debug/blazor.webassembly.js' does not exist. Enable NodeJS to build this project. [...src/aspnetcore/src/Components/WebAssembly/testassets/HostedInAspNet.Client/HostedInAspNet.Client.csproj]
...src/aspnetcore/src/Components/Directory.Build.targets(30,5): error : '...src/aspnetcore/src/Components/Web.JS/dist/Debug/blazor.webassembly.js' does not exist. Enable NodeJS to build this project. [...src/aspnetcore/src/Components/WebAssembly/testassets/StandaloneApp/StandaloneApp.csproj]
...src/aspnetcore/src/Components/Directory.Build.targets(30,5): error : '...src/aspnetcore/src/Components/Web.JS/dist/Debug/blazor.webassembly.js' does not exist. Enable NodeJS to build this project. [...src/aspnetcore/src/Components/benchmarkapps/Wasm.Performance/TestApp/Wasm.Performance.TestApp.csproj]
...src/aspnetcore/src/Components/Directory.Build.targets(30,5): error : '...src/aspnetcore/src/Components/Web.JS/dist/Debug/blazor.webassembly.js' does not exist. Enable NodeJS to build this project. [...src/aspnetcore/src/Components/WebAssembly/testassets/CustomBasePathApp/CustomBasePathApp.csproj]
...src/aspnetcore/src/Components/Directory.Build.targets(30,5): error : '...src/aspnetcore/src/Components/Web.JS/dist/Debug/blazor.webassembly.js' does not exist. Enable NodeJS to build this project. [...src/aspnetcore/src/Components/test/testassets/GlobalizationWasmApp/GlobalizationWasmApp.csproj]
...src/aspnetcore/src/Components/Directory.Build.targets(30,5): error : '...src/aspnetcore/src/Components/Web.JS/dist/Debug/blazor.webassembly.js' does not exist. Enable NodeJS to build this project. [...src/aspnetcore/src/Components/test/testassets/BasicTestApp/BasicTestApp.csproj]
...src/aspnetcore/src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj(104,5): error : '../../Web.JS/dist/Debug/blazor.server.js' does not exist. Falling back to checked-in copy.
...src/aspnetcore/src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj(104,5): error : '../../Web.JS/dist/Debug/blazor.server.js' does not exist. Falling back to checked-in copy.
...src/aspnetcore/src/Components/Directory.Build.targets(30,5): error : '...src/aspnetcore/src/Components/Web.JS/dist/Debug/blazor.webassembly.js' does not exist. Enable NodeJS to build this project. [...src/aspnetcore/src/Components/WebAssembly/testassets/Wasm.Authentication.Client/Wasm.Authentication.Client.csproj]
```

Add `-pack` and see one more (again, expected) error:
``` text
.../src/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Microsoft.AspNetCore.Components.WebAssembly.csproj(61,5): error : '../../../Web.JS/dist/Debug/blazor.webassembly.js' does not exist. Enable NodeJS to pack this project.
```

**Review by wtgodbe — APPROVED** [review]

File: AspNetCore.sln [hunk 2]
```diff
@@ -1060,7 +1060,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.CodeAnalysis.Razo
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "tools", "tools", "{F6E6A63A-45BE-46BF-B9B4-DA3DDC8FE4B5}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Razor.Internal.Transport", "src\Razor\tools\Microsoft.AspNetCore.Razor.Internal.Transport\Microsoft.AspNetCore.Razor.Internal.Transport.csproj", "{C3E588B7-AF20-448E-BAF0-D47C29B2E4B0}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Razor.Internal.Transport", "src\Razor\tools\Microsoft.Aspnetcore.Razor.Internal.Transport\Microsoft.AspNetCore.Razor.Internal.Transport.csproj", "{C3E588B7-AF20-448E-BAF0-D47C29B2E4B0}"
```
Comment (pranavkm) [inline AspNetCore.sln:1063]: Should we fix the casing of the project?

**Review by pranavkm — APPROVED** [review]

File: AspNetCore.sln (hunk 2 above)
Comment (dougbu) [reply AspNetCore.sln:1063]: It's actually a directory but I can change it instead if you like. Investigating the CI failures first&hellip;

File: build.sh [hunk 3]
```diff
@@ -237,6 +237,9 @@ elif [ "$build_all" != true ] && [ -z "$build_managed$build_nodejs$build_java$bu
     # We believe the most common thing our contributors will work on is C#, so if no other build group was picked, build the C# projects.
     __warn "No default group of projects was specified, so building the 'managed' and its dependent subset of projects. Run ``build.sh --help`` for more details."
     build_managed=true
+elif [ "$build_all" != true ] && [ -z "$build_managed" ] && ! [[ "$build_nodejs$build_java$build_native$build_installers" =~ "true" ]]; then
```
Comment (dougbu) [inline build.sh:240]: FYI the check above is a bit different than what we do in build.ps1. Here, we use one variable for `--build-xyz` and `--no-build-xyz` whereas Powershell keeps them separate 'til much later in the script. Adding this extra block makes things more similar.

Note builds can still fall back to attempting to build all of AspNetCore.sln e.g. `./build.ps1 --noBuildManaged` ☹️ Fortunately, that's a corner case.

**Comment by dougbu** [comment]
@dotnet/aspnet-build I'm very confused about the failing tests. Nothing looks related to this PR. Are those tests failing elsewhere❔

**Comment by dougbu** [comment]
For example,
``` text
System.ArgumentException : 'name' must not be null, empty or whitespace. (Parameter 'name')

Stack trace
   at Microsoft.Extensions.FileProviders.Embedded.Manifest.ManifestFile..ctor(String name, String resourcePath) in /_/src/FileProviders/Embedded/src/Manifest/ManifestFile.cs:line 16
   at Microsoft.Extensions.FileProviders.Embedded.Manifest.ManifestParser.BuildEntry(XElement element) in /_/src/FileProviders/Embedded/src/Manifest/ManifestParser.cs:line 90
   at Microsoft.Extensions.FileProviders.Embedded.Manifest.ManifestParser.BuildEntry(XElement element) in /_/src/FileProviders/Embedded/src/Manifest/ManifestParser.cs:line 110
   at Microsoft.Extensions.FileProviders.Embedded.Manifest.ManifestParser.Parse(Assembly assembly, String name) in /_/src/FileProviders/Embedded/src/Manifest/ManifestParser.cs:line 57
   at Microsoft.Extensions.FileProviders.Embedded.Manifest.ManifestParser.Parse(Assembly assembly) in /_/src/FileProviders/Embedded/src/Manifest/ManifestParser.cs:line 20
   at Microsoft.Extensions.FileProviders.ManifestEmbeddedFileProvider..ctor(Assembly assembly) in /_/src/FileProviders/Embedded/src/ManifestEmbeddedFileProvider.cs:line 26
   at Microsoft.AspNetCore.Components.Server.ConfigureStaticFilesOptions.PostConfigure(String name, StaticFileOptions options) in /_/src/Components/Server/src/DependencyInjection/ConfigureStaticFilesOptions.cs:line 61
   at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)
   at Microsoft.Extensions.Options.OptionsManager`1.<>c__DisplayClass5_0.<Get>b__0()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at Microsoft.Extensions.Options.OptionsCache`1.GetOrAdd(String name, Func`1 createOptions)
   at Microsoft.Extensions.Options.OptionsManager`1.Get(String name)
   at Microsoft.Extensions.Options.OptionsManager`1.get_Value()
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware..ctor(RequestDelegate next, IWebHostEnvironment hostingEnv, IOptions`1 options, ILoggerFactory loggerFactory) in /_/src/Middleware/StaticFiles/src/StaticFileMiddleware.cs:line 58
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.Internal.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider) in /_/src/Shared/ActivatorUtilities/ActivatorUtilities.cs:line 393
   at Microsoft.Extensions.Internal.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters) in /_/src/Shared/ActivatorUtilities/ActivatorUtilities.cs:line 87
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_0.<UseMiddleware>b__0(RequestDelegate next) in /_/src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs:line 96
   at Microsoft.AspNetCore.Builder.ApplicationBuilder.Build() in /_/src/Http/Http/src/Builder/ApplicationBuilder.cs:line 141
   at Microsoft.AspNetCore.Hosting.WebHost.BuildApplication() in /_/src/Hosting/Hosting/src/Internal/WebHost.cs:line 245
   at Microsoft.AspNetCore.Hosting.WebHost.StartAsync(CancellationToken cancellationToken) in /_/src/Hosting/Hosting/src/Internal/WebHost.cs:line 154
   at Microsoft.AspNetCore.TestHost.TestServer..ctor(IWebHostBuilder builder, IFeatureCollection featureCollection) in /_/src/Hosting/TestHost/src/TestServer.cs:line 97
   at Microsoft.AspNetCore.TestHost.TestServer..ctor(IWebHostBuilder builder) in /_/src/Hosting/TestHost/src/TestServer.cs:line 78
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateServer(IWebHostBuilder builder) in /_/src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs:line 385
...
```

**Comment by dougbu** [comment]
@dotnet/aspnet-blazor-eng or @dotnet/aspnet-build, any suggestions on resolving the Components.E2ETest and Mvc.FunctionalTests failures❔ I might be doing something wrong in this PR but it's not apparent to me. Sample stack trace [above](https://github.com/dotnet/aspnetcore/pull/29970#issuecomment-775596042).

**Comment by dougbu** [comment]
Still seeing Microsoft.AspNetCore.Components.E2ETests failures w/ the same `Exception` in https://dev.azure.com/dnceng/public/_build/results?buildId=988603

File: src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj [hunk 4]
```diff
@@ -91,11 +91,27 @@
...
+      <BlazorServerJSFile>..\..\Web.JS\dist\Release\blazor.server.js</BlazorServerJSFile>
+    </PropertyGroup>
+  </Target>
+
+  <Target Name="_AddEmbeddedBlazor" AfterTargets="_CheckBlazorServerJSPath">
+    <ItemGroup>
+      <EmbeddedResource Include="$(BlazorServerJSFile)" LogicalName="_framework/%(Filename)%(Extension)" />
+      <EmbeddedResource Include="$(BlazorServerJSFile).map" LogicalName="_framework/%(Filename)%(Extension)"
+          Condition="Exists('$(BlazorServerJSFile).map')" />
+    </ItemGroup>
```
Comment (dougbu) [inline src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj:115]: The problem leading to the test failures is in this target. I've downloaded a binary log to find the root cause.

When building in `main`, Microsoft.AspNetCore.Components.Server contains the following resources
image
and the embedded manifest contains
``` xml
<?xml version="1.0" encoding="utf-8" standalone="yes"?
<Manifest>
  <ManifestVersion>1.0</ManifestVersion>
  <FileSystem>
...
    <Directory Name="_framework">
      <File Name="blazor.server.js">
        <ResourcePath>_framework/blazor.server.js</ResourcePath>
      </File>
    </Directory>
  </FileSystem>
</Manifest>
```

But, the version created in this PR skips the blazor.server.js filename both in the resources
image
and the manifest
``` xml
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<Manifest>
  <ManifestVersion>1.0</ManifestVersion>
  <FileSystem>
...
    <Directory Name="_framework">
      <File Name="">
        <ResourcePath>_framework/</ResourcePath>
      </File>
    </Directory>
  </FileSystem>
</Manifest>
```

**Pull request merged**